Merge dev into feat-fast-sync

afck opened the pull request:
DO NOT MERGE!

This PR hopefully allows for an easier review of https://github.com/casper-network/casper-node/pull/2242: Instead of showing the diff with `feat-fast-sync` it shows the diff with `dev`.

Commit:
FEAT-44: Introducing fast sync

# Conflicts:
#	Makefile
#	client/lib/validation.rs
#	node/CHANGELOG.md
#	node/src/components/block_proposer.rs
#	node/src/components/block_validator/tests.rs
#	node/src/components/chainspec_loader.rs
#	node/src/components/contract_runtime.rs
#	node/src/components/fetcher.rs
#	node/src/components/fetcher/event.rs
#	node/src/components/linear_chain.rs
#	node/src/components/linear_chain/event.rs
#	node/src/components/linear_chain_sync.rs
#	node/src/components/linear_chain_sync/traits.rs
#	node/src/components/rpc_server/rpcs/state.rs
#	node/src/components/storage.rs
#	node/src/components/storage/tests.rs
#	node/src/effect.rs
#	node/src/effect/requests.rs
#	node/src/reactor/initializer.rs
#	node/src/reactor/joiner.rs
#	node/src/reactor/participating.rs
#	node/src/reactor/participating/error.rs
#	node/src/types.rs
#	node/src/types/block.rs

Commit:
Sunset legacy sync

  - Run deploys back to TTL (Closes #1596)

Commit:
Removing unecessary things from `LinearChainSync` component

afck committed:
Make Block::verify private.

Commit:
ChainspecLoader no longer running genesis/upgrades

afck committed:
Respond to all when fetching a block.

afck committed:
Fix cargo doc errors.

afck committed:
Replace TryInto with verify method.

afck committed:
Make UnverifiedBlock private.

afck committed:
Turn verify into Block::from_unverified.

Commit:
Merge pull request #1928 from afck/block_verify

Make Block::verify private.

Commit:
Genesis should not clobber an existing blockchain.

  - Closes #1937

Commit:
Fixing format

Commit:
Merge pull request #1938 from xcthulhu/genesis-no-clobber-existing-blockchain

Genesis should not clobber an existing blockchain

Commit:
Introducing `archival_sync`

Commit:
Verifying block from peer not-necessary (checked when constructed)

Commit:
Effect for getting blocks with sufficient finality signatures.

Commit:
Getting block and sufficient finality signatures by height

marc-casperlabs committed:
Use `MessageKind::StateTransfer` for transferred tries

Commit:
Simplifying how we download and store a block by hash.

# Conflicts:
#	node/src/components/linear_chain_sync/operations.rs

xcthulhu committed:
Clarifying CHANGELOG statement

Co-authored-by: Andreas Fackler <afck@users.noreply.github.com>

xcthulhu committed:
Getting rid of spurious comma

Co-authored-by: Andreas Fackler <afck@users.noreply.github.com>

xcthulhu committed:
Fixing done -> down typo

Co-authored-by: Andreas Fackler <afck@users.noreply.github.com>

xcthulhu committed:
Making comment for unused joiner event consistent with others

Co-authored-by: Andreas Fackler <afck@users.noreply.github.com>

marc-casperlabs committed:
Rename state transfers to trie transfers

Commit:
Merge pull request #1992 from marc-casperlabs/1644-message-kind-trie

Use `MessageKind::StateTransfer` for transferred tries

Commit:
Merge pull request #1998 from xcthulhu/sufficient-signatures

Sufficient Signatures

afck committed:
Remove EraSupervisorHandlingWrapper.

xcthulhu committed:
Chang wording in CHANGELOG around deprecated `starting_state_root_hash`

Co-authored-by: Fraser Hutchison <190532+Fraser999@users.noreply.github.com>

xcthulhu committed:
Chang wording around deprecated items in CHANGELOG

Co-authored-by: Fraser Hutchison <190532+Fraser999@users.noreply.github.com>

Commit:
Using era_start for ttl

Commit:
Reverting SIGNAL -> Exit code translation stuff

Commit:
Restoring `make lint`

Commit:
Removing unnecessary memory metrics

afck committed:
Fix era report/era end confusion.

Commit:
Fixing docstring reference to `ChainspecLoader`

Commit:
Adding changelog message to the appropriate section.

Commit:
block_fetcher -> block_by_hash_fetcher

Commit:
Removing [linear_chain_sync] section from various config tomls

Commit:
Removing spurious continue statement

Commit:
Removing unecessary loop label

Commit:
Getting rid of unused metrics

Commit:
Revert "Removing [linear_chain_sync] section from various config tomls"

Commit:
Disable cargo-audit from drone

Commit:
Removing linear_chain_sync from metrics

Commit:
Removing [linear_chain_sync] section from config.tomls

Commit:
Merge pull request #1917 from xcthulhu/sunset-legacy-sync

Sunset Legacy Sync

Commit:
Adding archival_sync field to [node] section in config tomls

Commit:
Fixing archival sync

Commit:
Introducing `trusted_hash_time_to_expiration` config option

  - Making archival_sync no longer optional

Commit:
If no block header is provided use the highest block from storage

Commit:
Introducing joining logic

Commit:
Do not run genesis on a node which is not version 1.0.0

xcthulhu committed:
Fixed grammar

Co-authored-by: Fraser Hutchison <190532+Fraser999@users.noreply.github.com>

xcthulhu committed:
Archival sync does not default to false.

Commit:
Better logging message

Commit:
Fixing formatting

Commit:
Reverting comment

xcthulhu committed:
Apply suggestions from code review

Co-authored-by: Andreas Fackler <afck@users.noreply.github.com>

Commit:
Ensuring that all have the genesis state after finishing archival sync

Commit:
Fixing formatting

Commit:
Using auction config and era duration to compute if a trusted hash is too old.

Commit:
Checking every block and state root has been downloaded

Commit:
Removing trusted_hash_time_to_expiration

Commit:
Fixed warning message

Commit:
Fixing lint errors

Commit:
Setting archival_sync to true

afck committed:
Only keep three eras in memory.

xcthulhu committed:
Changing logging message

Co-authored-by: Mateusz Górski <goral09@users.noreply.github.com>

xcthulhu committed:
More idiomatic use of the `info!` logging macro

Co-authored-by: Mateusz Górski <goral09@users.noreply.github.com>

Commit:
Only save fetcher information if it was retrieved from a peer

xcthulhu committed:
Get -> Gets

Co-authored-by: Andreas Fackler <afck@users.noreply.github.com>

Commit:
Only sync the trie in an upgrade if we can't find a new block.

xcthulhu committed:
Fixing doc comment

Commit:
Merge pull request #1995 from xcthulhu/archival-sync

Archival Sync

afck committed:
Add more comments about eras in memory.

afck committed:
Address review comments.

afck committed:
Remove consensus::ProtocolConfig.

afck committed:
Add era ID to error message.

afck committed:
Extract number of active eras into constant.

afck committed:
Add missed occurrence of PAST_ACTIVE_ERAS.

Co-authored-by: Mateusz Górski <goral09@users.noreply.github.com>

afck committed:
Replace hard-coded numbers with constants.

afck committed:
Call eras in memory "open", not "active".

The instantiated eras in memory include the evidence-only ones, that
don't have or accept any units anymore.

Commit:
Factored out `create_new_era` from `create_new_era_effects`

afck committed:
Make fast sync exit with codes to signal version upgrade or downgrade.

afck committed:
Fix build.

afck committed:
Add more sanity checks to create_new_era.

afck committed:
Make log messages lowercase.

Co-authored-by: Mateusz Górski <goral09@users.noreply.github.com>

afck committed:
Merge pull request #2127 from afck/eras_in_memory

Eras in memory

afck committed:
Deduplicate fetching blocks and block headers in fast sync.

afck committed:
Move version checks into fetch_and_store_by_block_height.

afck committed:
Add more checks to fetched blocks in fast sync.

afck committed:
Extract and unit test check_block_version.

afck committed:
Also check for downgrades when fetching the trusted key block.

afck committed:
Add a check for height overflow.

afck committed:
Disconnect from a peer that sent us a block with wrong parent.

afck committed:
Merge pull request #2180 from afck/exit_codes

Make fast sync exit with codes to signal version upgrade or downgrade.

afck committed:
Merge remote-tracking branch 'upstream/dev' into fast-sync-merge

afck committed:
Fix the should_fetch_from_peer test after merge.

The deploy acceptor now sets a deploy's `is_valid` flag, so the deploys
in the test must also have it for the equality test.

afck committed:
Fix run_equivocator_network test.

We keep fewer eras in memory now, and have a genesis block that is the
only block in era 0, so the test needs to be updated.

afck committed:
Fix RPC test; simplify fetcher test fix.

afck committed:
Remove chainspec from storage.

afck committed:
Bring contract runtime more in line with dev.

afck committed:
Reverting a few unneeded changes compared to dev.

afck committed:
Rename blake_hashes to hashes.

afck committed:
Merge remote-tracking branch 'upstream/dev' into fast-sync-merge

afck committed:
Apply rustfmt.

afck committed:
Apply right rustfmt version.

afck committed:
Apply rustfmt again.

afck committed:
Remove explicit compile_error.

afck committed:
Fix rustdoc.

afck committed:
Merge remote-tracking branch 'upstream/dev' into fast-sync-merge

afck committed:
Fix client no-std build.

afck committed:
Increase timeout for test_joiner_network.

It doesn't fail locally, so probably it's just slower in CI.

xcthulhu commented on client/lib/validation.rs:
```diff
@@ -13,7 +13,8 @@ use casper_node::{
         state::GlobalStateIdentifier,
```
As I have stated elsewhere, this file can be backported to `dev`.

afck replied:
Right, we'll have to revert https://github.com/casper-network/casper-node/pull/1928.

afck replied:
Done in https://github.com/casper-network/casper-node/pull/2266/commits/37798742bbe552ad6af2c64a41dc36bb47974017.

xcthulhu commented on node/src/components/block_validator.rs:
```diff
@@ -413,23 +413,44 @@ where
+                return Event::DeployMissing(dt_hash);
+            }
+        };
+        if deploy.deploy_or_transfer_hash() != dt_hash {
+            warn!(
+                deploy = ?deploy,
+                expected_deploy_or_transfer_hash = ?dt_hash,
+                actual_deploy_or_transfer_hash = ?deploy.deploy_or_transfer_hash(),
+                "Deploy has incorrect transfer hash"
+            );
+            return Event::CannotConvertDeploy(dt_hash);
         }
```
```suggestion
```

Fetcher handles this by design, so we can delete this.

afck replied:
Done in https://github.com/casper-network/casper-node/pull/2266/commits/ebb4cf508cd674d03abe767bd6bd0affb9418aac.

afck replied:
Turns out we do need this: The fetcher won't return a deploy with the wrong hash, but it doesn't distinguish between transfers and other deploys.

afck replied:
Reverted in https://github.com/casper-network/casper-node/pull/2266/commits/085c517cc6498958346c14c828122427e953f820.

xcthulhu replied:
🤦 my bad

xcthulhu commented on node/src/components/chainspec_loader.rs:
```diff
@@ -682,17 +346,7 @@ where
-                self.handle_upgrade_result(result);
-                Effects::new()
-            }
+            Event::Initialize => Effects::new(),
```
@marc-casperlabs The `initializer` reactor doesn't start if we don't have this initialize no-op event here.  Any ideas on how to remove it?

xcthulhu commented on node/CHANGELOG.md:
```diff
@@ -10,6 +10,7 @@ All notable changes to this project will be documented in this file.  The format
 [comment]: <> (Security:   in case of vulnerabilities)
 
 
+
```
```suggestion
```

afck replied:
Done in https://github.com/casper-network/casper-node/pull/2266/commits/ad57921ea9e92b6397cd300a427c6c4abb7b6bd1.

xcthulhu commented on node/src/components/consensus/config.rs:
```diff
@@ -45,46 +44,37 @@ impl Config {
-    pub(crate) genesis_timestamp: Option<Timestamp>,
-    /// The chainspec hash: All nodes in the network agree on it, and it's unique to this network.
-    pub(crate) chainspec_hash: Digest,
+pub trait ChainspecConsensusExt {
```
Since `Chainspec` is defined in this file, can't we just remove this and extend the inherent `impl Chainspec` instead?

afck replied:
But should we? These methods are quite consensus-specific and that would make them visible everywhere, without having to import anything. Not sure.

xcthulhu replied:
If we only want consensus to use this stuff, perhaps we should have this trait be `pub(super)`?

And if we want that, couldn't we accomplish the same thing with an inherent `impl` with `pub(super)` methods?

xcthulhu commented on node/src/components/consensus/consensus_protocol.rs:
```diff
@@ -9,14 +9,13 @@ use anyhow::Error;
 use datasize::DataSize;
 use serde::{Deserialize, Serialize};
```
This file should be reverted

afck replied:
Done in https://github.com/casper-network/casper-node/pull/2266/commits/ad57921ea9e92b6397cd300a427c6c4abb7b6bd1.

xcthulhu commented on node/src/components/consensus/era_supervisor.rs:
```diff
@@ -821,36 +707,29 @@ where
-    pub(super) fn handle_deactivate_era(
+    pub(super) fn handle_deactivate_era<REv: ReactorEventT<I>>(
         &mut self,
+        effect_builder: EffectBuilder<REv>,
```
These changes where we thread the effect builder and the prng everywhere *could* be backported, but it's probably a bunch of work compared to the other places identified.

xcthulhu commented on node/src/components/consensus/era_supervisor/era.rs:
```diff
@@ -55,11 +55,7 @@ pub struct Era<I> {
-    pub(crate) new_faulty: Vec<PublicKey>,
-    /// Validators that have been faulty in any of the recent BONDED_ERAS switch blocks. This
-    /// includes `new_faulty`.
+    /// Validators that have been faulty in any of the switch blocks after the booking block.
```
This is because we only need 2 eras now?

afck replied:
Actually it's because after that period they will be evicted anyway, and if they reactivate their bid we don't want to evict them again for the same past offense.

afck committed:
Fix minor style issues.

afck committed:
Revert #1928: Make Block::verify public again.

This approach to block verification won't be used after all. We might go
with the decorator approach instead later.

afck committed:
Add newline, so rustfmt doesn't change order.

xcthulhu commented on node/src/components/contract_runtime.rs:
```diff
@@ -629,22 +629,21 @@ impl ContractRuntime {
 
-        announcements::linear_chain_block(effect_builder, block, execution_results).await;
+        effect_builder
+            .announce_linear_chain_block(block, execution_results)
```
Why not just use the function in announcements?

afck replied:
Done in https://github.com/casper-network/casper-node/pull/2266/commits/9f0992d5a7044dfaa5c3060abeeb017fc8ede1bc.

(It's weird that the `ContractRuntimeAnnouncement` isn't in `effect::announcements` and doesn't use `EffectBuilder` methods like all the others, but I agree that change shouldn't be made on this branch.)

afck committed:
Revert removed newline.

xcthulhu commented on node/src/components/networking_metrics.rs:
```diff
@@ -32,6 +32,8 @@ pub(super) struct NetworkingMetrics {
     pub(super) out_count_deploy_transfer: IntCounter,
     /// Count of outgoing messages with block request/response payload.
     pub(super) out_count_block_transfer: IntCounter,
+    /// Count of outgoing messages with block request/response payload.
```
Should say trie not block in this comment

afck replied:
Done in https://github.com/casper-network/casper-node/pull/2266/commits/9f0992d5a7044dfaa5c3060abeeb017fc8ede1bc.

xcthulhu commented on node/src/components/consensus/consensus_protocol.rs:
```diff
@@ -163,7 +163,6 @@ impl<VID> EraReport<VID> {
         ])
     }
 }
-
```
No need to delete this line… keeps the diff smaller 😊

xcthulhu commented on node/src/components/contract_runtime.rs:
```diff
@@ -80,11 +81,6 @@ impl ExecutionPreState {
-
-    /// Get the next block height according that will succeed the block specified by `parent_hash`.
-    pub(crate) fn next_block_height(&self) -> u64 {
-        self.next_block_height
```
Do we still need next block height?  Isn’t it part of the ExecutionPreState?

afck replied:
It is! Looks like this function was never called, though.

afck committed:
Revert some accidental changes.

fizyk20 commented on node/CHANGELOG.md:
```diff
@@ -42,13 +44,22 @@ All notable changes to this project will be documented in this file.  The format
 * The `isolation_reconnect_delay` configuration, which has been ignored since 1.3, has been removed.
 * The libp2p-exclusive metrics of `read_futures_in_flight`, `read_futures_total`, `write_futures_in_flight`, `write_futures_total` have been removed.
+* Remove dead code revealed by making modules `pub(crate)`.
+* Remove legacy synchronization from genesis in favor of fast sync.
```
The other points are phrased like "X has been removed", should we rephrase these in this way to keep the style consistent?

afck replied:
Done in https://github.com/casper-network/casper-node/pull/2266/commits/c679d8151db4383a4e4ab8da3033ae501f40e1d3.

afck committed:
Address review comments.

fizyk20 commented on node/src/components/consensus/config.rs:
```diff
@@ -45,46 +44,37 @@ impl Config {
+    fn earliest_open_era(&self, current_era: EraId) -> EraId {
+        self.activation_era()
+            .successor()
+            .max(current_era.saturating_sub(2))
```
Should this `2` be hardcoded? We seem to be using a named constant `PAST_OPEN_ERAS` in consensus , if that's the same thing (and if not, another named constant would be preferable, I think).

afck replied:
Done in https://github.com/casper-network/casper-node/pull/2266/commits/c679d8151db4383a4e4ab8da3033ae501f40e1d3.

xcthulhu commented on node/src/components/fetcher/metrics.rs:
```diff
@@ -3,19 +3,19 @@ use prometheus::{IntCounter, Registry};
 
 #[derive(Debug)]
-pub(super) struct FetcherMetrics {
+pub struct FetcherMetrics {
```
Why can't this be `pub(super)` anymore?

afck replied:
It looks like a `metrics` method has been added to the `ItemFetcher` trait, which is `pub(crate)`.
But I can make them `pub(crate)`.

afck replied:
Done in https://github.com/casper-network/casper-node/pull/2266/commits/b937a91c50df68521d388955eef0cb1b0ae0706d.

fizyk20 commented on node/src/reactor/joiner.rs:
```diff
@@ -591,65 +630,103 @@ impl reactor::Reactor for Reactor {
-                        },
-                    };
-                    self.dispatch_event(
-                        effect_builder,
-                        rng,
-                        JoinerEvent::BlockByHeightFetcher(event),
-                    )
+                    match fetcher::Event::<BlockWithMetadata>::from_get_response_serialized_item(
+                        sender,
+                        &serialized_item,
+                    ) {
+                        Some(fetcher_event) => {
+                            self.dispatch_event(effect_builder, rng, JoinerEvent::BlockByHeightFetcher(fetcher_event))
+                        }
+                        None => {
+                            info!("{} sent us a block with metadata we couldn't parse! Banning", sender);
+                            effect_builder.announce_disconnect_from_peer(sender).ignore()
+                        }
+                    }
+                }
```
Didn't you already deduplicate this code in a clever way, or was that somewhere different? In any case, it would be nice to reduce the boilerplate here (with a macro, maybe?), but probably not in this PR - just as a general remark.

xcthulhu replied:
We deduplicated the "server" side in the `participating` reactor.  We can use a similar strategy here, I'm pretty sure we can get away without using a macro.

fizyk20 commented on node/src/reactor/participating.rs:
```diff
@@ -442,6 +434,102 @@ impl Reactor {
     }
 }
 
+impl Reactor {
+    /// Handles a request to get an item by id.
+    fn handle_get_request(
+        &self,
+        effect_builder: EffectBuilder<<Self as reactor::Reactor>::Event>,
+        sender: NodeId,
+        tag: Tag,
+        serialized_id: &[u8],
+    ) -> Effects<<Self as reactor::Reactor>::Event> {
```
Ah, here's the refactoring I had in mind in the comment I left in the joiner :sweat_smile: Might be nice to do something similar there (if possible - I see that these are requests, and I commented on the response handling in the joiner... anyway, not that important now).

afck replied:
I've actually already started doing that a while ago. But I'll probably have to start over after `dev` is merged into `feat-fast-sync`. :grimacing:

fizyk20 commented on node/src/reactor/participating/config.rs:
```diff
@@ -4,16 +4,15 @@ use serde::Deserialize;
 #[serde(deny_unknown_fields)]
 pub(crate) struct Config {
-    /// Node configuration.
+    /// Node fast-sync configuration.
```
Is that only the fast sync config now? If so, maybe we should rename the struct?

xcthulhu replied:
I think @Fraser999 had some pattern he wanted

xcthulhu commented on node/src/components/storage.rs:
```diff
@@ -72,25 +68,23 @@ use casper_types::{
 };
 
+pub(crate) use crate::effect::requests::StorageRequest; /* Needed by reactor! macro in
+                                                         * fetcher tests... */
```
```suggestion
// The reactor! macro needs this in the fetcher tests
#[cfg(test)]
pub(crate) use crate::effect::requests::StorageRequest;
```

afck replied:
Done in https://github.com/casper-network/casper-node/pull/2266/commits/c679d8151db4383a4e4ab8da3033ae501f40e1d3.

xcthulhu commented on node/src/components/storage.rs:
```diff
@@ -184,6 +167,7 @@ pub enum Error {
     /// LMDB error while operating.
     #[error("internal database error: {0}")]
     InternalStorage(#[from] LmdbExtError),
+
```
```suggestion
```

afck replied:
Done in https://github.com/casper-network/casper-node/pull/2266/commits/c679d8151db4383a4e4ab8da3033ae501f40e1d3.

xcthulhu commented on node/src/components/storage.rs:
```diff
@@ -461,6 +447,7 @@ impl Storage {
 
         let mut deleted_block_hashes = HashSet::new();
         let mut deleted_block_body_hashes_v1 = HashSet::new();
+        let mut any_v2_block_deleted = false;
```
@fizyk20 didn't we backport this to `dev`?

fizyk20 replied:
Ah, that's because the changes on feat-fast-sync seem to be older. When we removed the call to the garbage collection, the indication whether any blocks were removed stopped being necessary. So the version that is on `dev` is actually the newer one! So we should keep this (and `initialize_block_body_v2_db`) the way it is on `dev`. cc @afck

afck replied:
Done in https://github.com/casper-network/casper-node/pull/2266/commits/bbe7d203e46091eaf75ea7fd0a2a6b997da807b5. Please double-check that I reverted the right things.

fizyk20 replied:
Looks right to me :+1: In case you'd like to double-check, these changes were made in #1934

afck committed:
Address review comments.

afck committed:
Make FetcherMetrics pub(crate).

afck committed:
Add back the deploy hash check.

This is needed after all: The fetcher won't return a deploy with the
wrong hash, but it doesn't distinguish between transfers and other
deploys.

afck committed:
Use storage integrity check code from dev.

afck committed:
Fix formatting.

xcthulhu commented on node/CHANGELOG.md:
```diff
@@ -42,13 +44,22 @@ All notable changes to this project will be documented in this file.  The format
 * Experimental support for libp2p has been removed.
 * The `isolation_reconnect_delay` configuration, which has been ignored since 1.3, has been removed.
 * The libp2p-exclusive metrics of `read_futures_in_flight`, `read_futures_total`, `write_futures_in_flight`, `write_futures_total` have been removed.
+* Dead code revealed by making modules `pub(crate)` has been removed.
```
This already happened on `dev`, so it looks like that changelog message got clobbered upstream.  Maybe remove?

afck replied:
Done in https://github.com/casper-network/casper-node/pull/2266/commits/edc6424924de5a4bae1b3365804e8d9ccab75dc3.

xcthulhu commented on node/src/components/storage.rs:
```diff
@@ -809,15 +732,36 @@ impl Storage {
+                           actual_block_hash = ?block.hash(),
+                           "block not stored under hash");
+                    debug_assert_eq!(&block_hash, block.hash());
+                }
```
```suggestion
                    return responder.respond(None).ignore();
                }
```

afck replied:
Good catch!

afck replied:
Done in https://github.com/casper-network/casper-node/pull/2266/commits/edc6424924de5a4bae1b3365804e8d9ccab75dc3.

xcthulhu commented on node/src/components/storage.rs:
```diff
@@ -809,15 +732,36 @@ impl Storage {
+                if finality_signatures.verify().is_err() {
+                    error!(?block, "invalid finality signatures for block");
+                    debug_assert!(finality_signatures.verify().is_ok());
+                }
```
```suggestion
                    return responder.respond(None).ignore();
                }
```

afck replied:
Done in https://github.com/casper-network/casper-node/pull/2266/commits/edc6424924de5a4bae1b3365804e8d9ccab75dc3.

xcthulhu commented on node/src/components/storage.rs:
```diff
@@ -709,7 +629,10 @@ impl Storage {
                 .respond(self.get_transfers(&mut self.env.begin_ro_txn()?, &block_hash)?)
                 .ignore(),
             StorageRequest::PutDeploy { deploy, responder } => {
-                responder.respond(self.put_deploy(&*deploy)?).ignore()
```
Hmm... what happened to `put_deploy`?  Maybe we should just leave this in to minimize the diff?

afck replied:
Done in https://github.com/casper-network/casper-node/pull/2266/commits/edc6424924de5a4bae1b3365804e8d9ccab75dc3.

afck committed:
Address comments; storage fixes.

xcthulhu commented on node/src/effect/requests.rs:
```diff
@@ -847,30 +810,32 @@ impl Display for ContractRuntimeRequest {
 
 /// Fetcher related requests.
 #[derive(Debug, Serialize)]
 #[must_use]
```
`#[must_use]`?

afck replied:
I guess the idea is that you shouldn't drop something that contains a responder.
Anyway, seems to be the same on `dev`?

xcthulhu replied:
Just wondering, thanks for the explanation...

xcthulhu commented on node/src/reactor/participating/tests.rs:
```diff
@@ -240,6 +240,7 @@ impl SwitchBlocks {
         &self.headers[era_number as usize]
             .era_end()
             .expect("era end")
+            .era_report()
```
Maybe backport this to `dev`?

afck replied:
Good idea! I added it to https://github.com/casper-network/casper-node/issues/2274.

xcthulhu commented on node/src/testing/test_rng.rs:
```diff
@@ -24,6 +24,7 @@ type Seed = <Pcg64Mcg as SeedableRng>::Seed; // [u8; 16]
 /// thread in which it is created panics.
 ///
 /// Only one `TestRng` is permitted per thread.
+#[derive(Clone)]
```
When do we clone this?

xcthulhu commented on node/src/types/chainspec.rs:
```diff
@@ -47,7 +47,7 @@ pub const CHAINSPEC_NAME: &str = "chainspec.toml";
 /// A collection of configuration settings describing the state of the system at genesis and after
 /// upgrades to basic system functionality occurring after genesis.
-#[derive(DataSize, PartialEq, Eq, Serialize, Debug)]
+#[derive(DataSize, PartialEq, Eq, Serialize, Debug, Clone)]
```
Since we always have one behind an `Arc` in practice, maybe we don't need `Clone` here?

xcthulhu commented on node/src/types/deploy.rs:
```diff
@@ -935,6 +935,25 @@ impl Deploy {
         self.header.chain_name.clear();
     }
 
+    /// Returns `true` if the two deploys are equal except for possibly the `is_valid` flag. I.e.
```
🤮

xcthulhu replied:
Maybe just impl `PartialEq` manually?

afck replied:
I wasn't sure if it's a good idea to change the `Eq` implementation itself.
Another option is to validate all the deploys in the tests, so that the field does have the same value.

xcthulhu replied:
> Another option is to validate all the deploys in the tests, so that the field does have the same value.

I like that.

Don't we have a ticket to introduce a `Validated<T>` decorator? We can move to that and get rid of this eventually...

xcthulhu commented on node/src/types/exit_code.rs:
```diff
@@ -9,7 +10,7 @@ const SIGNAL_OFFSET: u8 = 128;
 ///
 /// Note that a panic will result in the Rust process producing an exit code of 101.
-#[derive(Clone, Copy, PartialEq, Eq, Debug, DataSize)]
+#[derive(Clone, Copy, PartialEq, Eq, Debug, DataSize, Serialize, Deserialize)]
```
```suggestion
#[derive(Clone, Copy, PartialEq, Eq, Debug, DataSize)]
```

afck replied:
You're right, I'll add `#[serde(skip_serializing)]` to the `JoinerEvent` variant. I'm not entirely sure why that type needs to be serializable in the first place.

xcthulhu commented on node/src/types/status_feed.rs:
```diff
@@ -164,7 +154,8 @@ pub struct GetStatusResult {
     /// The state root hash used at the start of the current session.
-    pub starting_state_root_hash: Digest,
+    #[deprecated(since = "1.5.0")]
+    pub starting_state_root_hash: Option<Digest>,
```
```suggestion
    pub starting_state_root_hash: Digest,
```

afck replied:
So we're not deprecating it after all?

xcthulhu replied:
It's still being deprecated, but we can't change the RPC API.  Putting a `#[deprecated(since = "1.5.0")]` makes the compiler throw a warning if you try to access this value.  Below we should return a sentinel when we must give a value.  I suggested `[0u8; 32]`, but we could also use`[0xde, 0xad, 0xbe, 0xef, 0xde, 0xad, 0xbe, ...]` or something.

The `#[deprecated(since = "1.5.0")]` annotation also updates the JSON Schema accordingly.

afck replied:
Done in https://github.com/casper-network/casper-node/pull/2266/commits/e63773f50348ce3ba37bb9c24466366e64f863ad.

xcthulhu commented on node/src/types/status_feed.rs:
```diff
@@ -182,11 +173,12 @@ pub struct GetStatusResult {
             api_version,
             chainspec_name: status_feed.chainspec_info.name,
-            starting_state_root_hash: status_feed.chainspec_info.starting_state_root_hash,
+            starting_state_root_hash: None,
```
```suggestion
            starting_state_root_hash: Digest::from([0u8 ; 32])
```

afck replied:
Done in https://github.com/casper-network/casper-node/pull/2266/commits/e63773f50348ce3ba37bb9c24466366e64f863ad.

afck committed:
Remove unneeded Clone/Serialize derives.

xcthulhu commented on smart_contracts/contract/src/lib.rs:
```diff
@@ -57,6 +57,7 @@
     test(attr(forbid(warnings)))
 )]
 #![warn(missing_docs)]
+
```
```suggestion
```

afck replied:
I'd keep that: The attribute applies to whole crate (note the `!`!), not the next line.

xcthulhu commented on utils/retrieve-state/src/storage.rs:
```diff
@@ -135,6 +137,8 @@ pub fn create_storage(chain_download_path: impl AsRef<Path>) -> Result<Storage,
         ProtocolVersion::from_parts(0, 0, 0),
         false,
         "test",
+        Ratio::new(1, 3),
```
👍 

@dwerner this is basically the only value you'll ever want in practice.

afck committed:
Revert API change.

afck commented:
@xcthulhu: Please approve #2242 instead; this one here is just for the review.

afck commented:
https://github.com/casper-network/casper-node/pull/2242 has been merged.
